add readme parsing tests

SaketramDurbha opened the pull request:
This PR addresses issue #5 .

- Add readme parsing tests
- Refactor acutal readme parsing code to be more testable

Will rebase and update once PR #22 is merged.

Commit:
add readme tests (first attempt)

Commit:
refactor readme parsing code and tests
- add more test cases
- make code more testable

Commit:
minor updates to test error messages

Commit:
add readme parsing tests

Commit:
fix typo in lifecycle/readme.go

kurtisvg commented on internal/lifecycle/readme.go:
```diff
@@ -42,7 +42,11 @@ var (
 	mdCodeFenceStartRegexp = regexp.MustCompile("^\\w*`{3,}[^`]*$")
 
-	errNoREADMECodeBlocksFound = fmt.Errorf("[lifecycle.parseREADME]: no code blocks immediately preceded by %s found", codeTag)
+	errNoREADMECodeBlocksFound   = fmt.Errorf("[lifecycle.parseREADME]: no code blocks immediately preceded by %s found", codeTag)
```
We drop the `[funcName]` in another PR I think - should we do so here as well?

kurtisvg replied:
if these errors aren't being returned in multiple locations, It's probably not worth declaring in advance.

SaketramDurbha replied:
Resolved offline, thank you.

kurtisvg commented on internal/lifecycle/readme.go:
```diff
@@ -109,9 +113,19 @@ func parseREADME(filename, serviceName, gcrURL string) (Lifecycle, error) {
+	return l, nil
+}
+
+func extractLifecycle(scanner *bufio.Scanner, serviceName, gcrURL string) (Lifecycle, error) {
```
Should you add a comment to this function?

SaketramDurbha replied:
Resolved, thank you.

kurtisvg commented on internal/lifecycle/readme_test.go:
```diff
@@ -0,0 +1,438 @@
+			return err
+		}
+	}
+
```
nit: remove newline

SaketramDurbha replied:
Resolved, thank you.

kurtisvg commented on internal/lifecycle/readme_test.go:
```diff
@@ -0,0 +1,438 @@
+		err := os.Setenv(k, v)
+		if err != nil {
+			return err
+		}
```
nit:
```suggestion
		if err := os.Setenv(k, v); err != nil {
			return err
		}
```

SaketramDurbha replied:
Resolved, thank you.

kurtisvg commented on internal/lifecycle/readme_test.go:
```diff
@@ -0,0 +1,438 @@
+	"testing"
+)
+
+func setEnv(e map[string]string) error {
```
nit: Comments are good, even on utility functions like this.

SaketramDurbha replied:
Resolved, thank you.

kurtisvg commented on internal/lifecycle/readme_test.go:
```diff
@@ -0,0 +1,438 @@
+	if b == nil {
+		return a == nil
+	}
+	return errors.Is(a, b)
```
hmm, I'm not sure what this is doing, but it seems iffy. What are you trying to do?

SaketramDurbha replied:
Resolved, thank you.

kurtisvg commented on internal/lifecycle/readme_test.go:
```diff
@@ -0,0 +1,438 @@
+
+var tests = []test{
+	{
+		in: "[//]: # ({sst-run-unix})\n" +
```
Consider multiline strings or a single line - this feels a bit hard to read.

SaketramDurbha replied:
Resolved offline, thank you.

SaketramDurbha replied:
For reference: we're keeping this because it doesn't seem line multiline strings support escaped backticks.

kurtisvg commented on internal/lifecycle/readme_test.go:
```diff
@@ -0,0 +1,438 @@
+
+		h, err := extractCodeBlocks(s)
+		if !equalError(err, tc.extractCodeBlocksErr) {
+			t.Errorf("#%d: error mismatch\nhave: %v\nwant: %v", i, err, tc.extractCodeBlocksErr)
```
Go has a common idiom for this: `want x, got y`

SaketramDurbha replied:
Resolved offline, thank you.

SaketramDurbha replied:
Resolved offline, thank you.

kurtisvg commented on internal/lifecycle/readme_test.go:
```diff
@@ -0,0 +1,438 @@
+}
+
+var tests = []test{
+	{
```
I would consider adding a comment to each of these - what are these parameters supposed to be testing?

SaketramDurbha replied:
Resolved, thank you.

Commit:
address first round of PR reviews, code clean up

- comment lifecycle.extractLifecycle
- add comments to each test in readme_test.go
- update mismatch error message format to readme_test.go
- remove unnecessary equalError function in readme_test.go
- other minor code clean up

Commit:
add actual file parsing test to readme parsing tests

Commit:
refactor file parsing test in readme parsing tests

kurtisvg commented on internal/lifecycle/readme_test.go:
```diff
@@ -0,0 +1,504 @@
+}
+
+func TestToCommands(t *testing.T) {
+	for i, tc := range tests {
```
Rather than iterate through each list of tests for each section, you should probably just have a different slice for each test.

SaketramDurbha replied:
Resolved, thank you.

Commit:
refactor readme_test.go

kurtisvg commented on internal/lifecycle/readme_test.go:
```diff
@@ -0,0 +1,448 @@
+	}
+}
+
+type parseREADMETest struct {
```
nit: Use MixedCaps

SaketramDurbha replied:
Resolved, thank you.

kurtisvg commented on internal/lifecycle/readme_test.go:
```diff
@@ -0,0 +1,448 @@
+			continue
+		}
+
+		if err == nil && !reflect.DeepEqual(lifecycle, tc.lifecycle) {
```
What if err != nil?

SaketramDurbha replied:
I wasn't sure if we wanted to still check the results if there was an error.

SaketramDurbha replied:
After talking offline, we're leaving this as is. Thank you.

kurtisvg reviewed (approved):
2 nits, then LGTM

Commit:
minor readme code refactor
- TestParseReadme code fix
- README -> Readme in code

dmahugh reviewed (approved):
LGTM

tritone commented on internal/lifecycle/lifecycle.go:
```diff
@@ -58,22 +58,20 @@ func NewLifecycle(sampleDir, serviceName, gcrURL string) (Lifecycle, error) {
-
-
 	if _, err := os.Stat(readmePath); err == nil {
-		lifecycle, err := parseREADME(readmePath, serviceName, gcrURL)
```
All caps on README is actually appropriate here I think.

SaketramDurbha replied:
Resolved, thank you.

tritone commented on internal/lifecycle/readme.go:
```diff
@@ -42,7 +42,11 @@ var (
 	mdCodeFenceStartRegexp = regexp.MustCompile("^\\w*`{3,}[^`]*$")
 
-	errNoREADMECodeBlocksFound = fmt.Errorf("lifecycle.extractCodeBlocks: no code blocks immediately preceded by %s found", codeTag)
+	errNoReadmeCodeBlocksFound   = fmt.Errorf("lifecycle.extractCodeBlocks: no code blocks immediately preceded by %s found", codeTag)
```
I actually don't think this change is necessary, unless any of these errors are going to be used in more than one place. It makes it a little more complicated to read the code.

If anything, it might be easier for callers if there was a single error type that wrapped any error coming from readme parsing.

SaketramDurbha replied:
Resolved offline, thank you.

tritone commented on internal/lifecycle/readme.go:
```diff
@@ -68,7 +72,7 @@ func (cb codeBlock) toCommands(serviceName, gcrURL string) ([]*exec.Cmd, error)
 			i++
 			if i >= len(cb) {
-				return nil, fmt.Errorf("unexpected end of code block: expecting command line continuation; code block dump:\n%s", strings.Join(cb, "\n"))
+				return nil, fmt.Errorf("%w; code block dump:\n%s", errCodeBlockEndAfterLineCont, strings.Join(cb, "\n"))
```
The use of wrapping here doesn't really make sense-- usually that would just be when you are bubbling up an error from a called func.

SaketramDurbha replied:
Resolved, thank you.

tritone commented on internal/lifecycle/readme.go:
```diff
@@ -204,15 +218,15 @@ func replaceServiceName(command, serviceName string) string {
 
 	// Searches for specific gcloud keywords and takes service name from them
-	for i := 0; i < len(sp) - 1; i++ {
+	for i := 0; i < len(sp)-1; i++ {
```
Were these formatting changes yours or from gofmt?

SaketramDurbha replied:
`gofmt`.

tritone commented on internal/lifecycle/readme_test.go:
```diff
@@ -0,0 +1,452 @@
+}
+
+func TestToCommands(t *testing.T) {
+	for i, tc := range toCommandsTests {
```
Having a list of test cases and then looping through them seems like the right approach. I would also consider using a subtest for each case, which produces nicer output. This seems like it could work for most of your tests.

https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks
https://blog.golang.org/subtests

SaketramDurbha replied:
Creating an issue for this.

SaketramDurbha replied:
Actually, a PR: https://github.com/GoogleCloudPlatform/serverless-sample-tester/pull/31

tritone commented on internal/lifecycle/readme_test.go:
```diff
@@ -0,0 +1,452 @@
+			"echo multi \\",
+		},
+		cmds: nil,
+		err:  errCodeBlockEndAfterLineCont,
```
In general it's better not to have your test use unexported stuff from the package. Think about what kind of error a caller outside the package would expect to see. String matching is probably fine, or you could think about making an exported error that callers could actually check for.

SaketramDurbha replied:
Resolved offline, thank you.

Commit:
revert README -> Readme in code

Commit:
don't wrap errors in codeBlock.toCommands
